File: local-movie-app/src/main/java/com/github/rahmnathan/localmovies/app/control/MovieEventLoader.java
package com.github.rahmnathan.localmovies.app.control;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import com.github.rahmnathan.localmovies.app.activity.SetupActivity;
import com.github.rahmnathan.localmovies.app.adapter.list.MovieListAdapter;
import com.github.rahmnathan.localmovies.app.data.Client;
import com.github.rahmnathan.localmovies.app.adapter.external.localmovie.MovieFacade;
import com.github.rahmnathan.localmovies.app.data.Movie;
import com.github.rahmnathan.localmovies.app.data.MovieEvent;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.Collectors;

public class MovieEventLoader implements Runnable {
    private final Logger logger = Logger.getLogger(MovieEventLoader.class.getName());
    private final Handler UIHandler = new Handler(Looper.getMainLooper());
    private final MoviePersistenceManager persistenceManager;
    private final MovieFacade movieFacade = new MovieFacade();
    private final MovieListAdapter movieListAdapter;
    private final Context context;
    private final Client client;

    public MovieEventLoader(MovieListAdapter movieListAdapter, Client myClient, MoviePersistenceManager persistenceManager, Context context) {
        this.movieListAdapter = movieListAdapter;
        this.persistenceManager = persistenceManager;
        this.client = myClient;
        this.context = context;
    }


    @Override
    public void run() {
        logger.log(Level.INFO, "Dynamically loading titles");

        if (client.getAccessToken() == null) {
            UIHandler.post(() -> Toast.makeText(context, "Login failed - Check credentials", Toast.LENGTH_LONG).show());
            return;
        }

        List<MovieEvent> events = movieFacade.getMovieEvents(client);
        events.forEach(event -> {
            if(event.getEvent().equalsIgnoreCase("CREATE")){
                Movie movie = event.getMovie();
                persistenceManager.addOne(getParentPath(event.getRelativePath()), movie);
                movieListAdapter.clearLists();
                movieListAdapter.updateList(persistenceManager.getMoviesAtPath(client.getCurrentPath().toString()).orElse(new ArrayList<>()));
                UIHandler.post(movieListAdapter::notifyDataSetChanged);
            } else {
                persistenceManager.deleteMovie(event.getRelativePath());
                movieListAdapter.clearLists();
                movieListAdapter.updateList(persistenceManager.getMoviesAtPath(client.getCurrentPath().toString()).orElse(new ArrayList<>()));
                UIHandler.post(movieListAdapter::notifyDataSetChanged);
            }
        });

        if (!movieListAdapter.getChars().equals("")) {
            UIHandler.post(() -> movieListAdapter.getFilter().filter(movieListAdapter.getChars()));
        }

        client.setLastUpdate(System.currentTimeMillis());
        SetupActivity.saveData(client, context);
    }

    private static String getParentPath(String path){
        String[] dirs = path.split(File.separator);
        return Arrays.stream(dirs)
                .limit(dirs.length - 1)
                .collect(Collectors.joining(File.separator));
    }
}

File: local-movie-app/src/main/java/com/github/rahmnathan/localmovies/app/control/MovieEventLoader.kt
package com.github.rahmnathan.localmovies.app.control

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import com.github.rahmnathan.localmovies.app.activity.SetupActivity.Companion.saveData
import com.github.rahmnathan.localmovies.app.adapter.external.localmovie.MovieFacade.getMovieEventCount
import com.github.rahmnathan.localmovies.app.adapter.external.localmovie.MovieFacade.getMovieEvents
import com.github.rahmnathan.localmovies.app.adapter.list.MovieListAdapter
import com.github.rahmnathan.localmovies.app.control.MediaPathUtils.getParentPath
import com.github.rahmnathan.localmovies.app.data.Client
import com.github.rahmnathan.localmovies.app.data.MovieEvent
import java.util.*
import java.util.function.Consumer
import java.util.logging.Level
import java.util.logging.Logger

class MovieEventLoader(private val movieListAdapter: MovieListAdapter, private val client: Client, private val persistenceManager: MoviePersistenceManager, private val context: Context) : Runnable {
    private val logger = Logger.getLogger(MovieEventLoader::class.java.name)
    private val UIHandler = Handler(Looper.getMainLooper())

    override fun run() {
        logger.log(Level.INFO, "Dynamically loading events.")
        if (client.accessToken == null) {
            UIHandler.post { Toast.makeText(context, "Login failed - Check credentials", Toast.LENGTH_LONG).show() }
            return
        }
        val count = getMovieEventCount(client)
        if (!count.isPresent) {
            logger.severe("Error retrieving media event count.")
            return
        }

        for (page in 0..count.get() / ITEMS_PER_PAGE) {
            val events = getMovieEvents(client, page.toInt(), ITEMS_PER_PAGE)
            events.forEach(Consumer { event: MovieEvent ->
                logger.info("Found media event: $event")
                if (event.event.equals("CREATE", ignoreCase = true)) {
                    val media = event.media
                    persistenceManager.deleteMovie(event.relativePath)
                    persistenceManager.addOne(getParentPath(event.relativePath), media!!)
                    movieListAdapter.clearLists()
                    movieListAdapter.updateList(persistenceManager.getMoviesAtPath(client.currentPath.toString()).orElse(ArrayList()))
                    UIHandler.post { movieListAdapter.notifyDataSetChanged() }
                } else {
                    persistenceManager.deleteMovie(event.relativePath)
                    movieListAdapter.clearLists()
                    movieListAdapter.updateList(persistenceManager.getMoviesAtPath(client.currentPath.toString()).orElse(ArrayList()))
                    UIHandler.post { movieListAdapter.notifyDataSetChanged() }
                }
            })
        }

        if (movieListAdapter.chars != "") {
            UIHandler.post { movieListAdapter.filter.filter(movieListAdapter.chars) }
        }

        client.lastUpdate = System.currentTimeMillis()
        saveData(client, context)
    }

    companion object {
        private const val ITEMS_PER_PAGE = 30
    }

}